ProjectService_DbContext

mekk1t opened the pull request:
A new project is added to the solution. 
DbContext is defined and the database is created locally using PM console.

mekk1t committed:
Initial Create: ProjectService

Added a new empty Web Application project "ProjectService" to the "DigitalOffice" solution.

mekk1t committed:
DbContext and Entity definition

Created a new folder "Database" where i defined DbContext, creating a new class "ProjectServiceDbContext". In there I defined the table it'll be using -- Projects.

In the "Database" folder created a new one, called "Entities". There I created a class that defines the table "Projects". Called it "Project.cs".

In "Project.cs" I used two attributes: [Key] and [NotMapped]
Key is for the primary key definition
NotMapped is for excluding the property from the table.
I got confused with the implementation of these fields because we're working with the separated databases.

mekk1t committed:
DbContext registration

Added a "ConnectionStrings" section in the "appsettings.json" file.
It contains a first entry called "SQLConnectionString" that is used for connecting to the SQL Server.

In "Startup.cs" I added "Configuration" field so that I can access the configuration file.
DbContext is registered via "services.AddDbContext".

mekk1t committed:
First DbContext Migration

Using PM console, typed:
1. Add-Migration InitialCreate
2. Update-Database -Migration InitialCreate.

It created a new folder "Migrations", captured all the definitions and then updated the database. In this case, database did not exist, so EntityFrameworkCore created it using the provided in config file "SQLConnectionString".

denis-kriv commented on DigitalOffice/ProjectService/Startup.cs:
```diff
@@ -0,0 +1,45 @@
+            services.AddControllers();
+        }
+        
+        public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
+        {
+            if (env.IsDevelopment())
+            {
+                app.UseDeveloperExceptionPage();
+            }
+
+            app.UseRouting();
+
+            app.UseEndpoints(endpoints =>
+            {
+                endpoints.MapControllers();
+            });
+        }
```
How about `app.UseHttpsRedirection();` ?

mekk1t replied:
@DenisKrivoruchko11 could you please explain why you think we should use it?
I googled and it's used when the app is in production. While we're developing, I think it's not necessary.

ifelseelif replied:
How about  'The structure should be like that: .../Services/ProjectService' ?

mekk1t replied:
@ifelseelif @djino97 @DenisKrivoruchko11 @KostinP @lalagvana 
I think we can discuss the structure on the next call at 15.30

denis-kriv reviewed (commented):
The structure should be like that: .../Services/ProjectService

djino97 commented:
I suggest naming files related to description of models DB, example: ProjectModels (in Entities)

mekk1t commented:
@djino97 didnt understand your suggestion.
You suggest that we rename "Project.cs" into "ProjectModel.cs"?

By the way, @ifelseelif @lalagvana @KostinP 
We should consider naming our Database Models like "DbModel.cs", e.g. "DbProject.cs" or we could go with "Project.cs"
Thoughts? Either we name it as in this Pull Request or later we will call front-end models something like "ModelDto.cs"

mekk1t committed:
Updated ports

Ports are changed from 5000 to different ones so that in the future we'll be able to run multiple projects simultaneously

mekk1t committed:
New folder "Services", new localhost ports, added .UseHttpRedirection

Moved ProjectService files to a new folder "Services"

----
Also updated ports and added .UseHttpRedirection() to Startup.cs

mekk1t committed:
Updated Entities and added 2 new ones

Created 2 new models:
ProjectWorkerUser.cs and ProjectManagerUser.cs that serve as a List<Guid> for the Projects table.
These new tables contain pairs of ProjectId and WorkerUserId/ManagerUserId

Migrations are applied and the database is updated. Everything's in order.

mekk1t committed:
Automated Migrations

Migrations are applied to the addressed DbContext in a new private static method that gets a scope of the dbContext and updates its database.

mekk1t committed:
Added ProjectFile Entity

Created a new entity called "ProjectFile" that serves as a link between Projects and its Files.
That table contains pairs of Ids -- ProjectId + FileId.

Migration created, applied to the database.

Co-Authored-By: Evgeniy <38541098+djino97@users.noreply.github.com>

mekk1t committed:
Obsolete property --> IsActive property

Renamed the "disabling" property

Co-Authored-By: Evgeniy <38541098+djino97@users.noreply.github.com>

nmamro commented on DigitalOffice/Services/ProjectService/Database/Entities/ProjectWorkerUser.cs:
```diff
@@ -0,0 +1,12 @@
+        public Guid ProjectId { get; set; }
+        public Project Project { get; set; }
+        public Guid WorkerUserId { get; set; }
+
```
Formatting is kinda good thing, empty lines are bruh

nmamro commented on DigitalOffice/Services/ProjectService/Database/Entities/ProjectManagerUser.cs:
```diff
@@ -0,0 +1,11 @@
+
+namespace ProjectService.Database.Entities
+{
+    public class ProjectManagerUser
```
Idk if "user" suffix is necessary. Or is it?

mekk1t replied:
Yes it is. We discussed this on our previous call: we mark our properties (and models, if as in this case they serve as a link between microservices) so that we understand where does this data service come from.

nmamro commented on DigitalOffice/Services/ProjectService/Startup.cs:
```diff
@@ -0,0 +1,62 @@
+            });
+        }
+
+        private static void UpdateDatabase(IApplicationBuilder app)
```
Why static btw?

mekk1t replied:
You can explain why this shouldn't be static?

As StackOverflow states, "static" allows the compiler to skip implicit "this." pointer and we make sure that this method can't change Startup's object state, etc.

Also this method will be called each time we start the application, so even a tiny performance gain would be good, I guess.

nmamro replied:
> You can explain why this shouldn't be static?
> 
> As StackOverflow states, "static" allows the compiler to skip implicit "this." pointer and we make sure that this method can't change Startup's object state, etc.
> 
> Also this method will be called each time we start the application, so even a tiny performance gain would be good, I guess.

I didn't suggest any changes, just asked ;)

mekk1t committed:
Update DigitalOffice.sln

Added a solution folder "Services", moved ProjectService there

mekk1t committed:
Deleted "Services" Folder

Moved ProjectService files from "Services" folder to "Services" SOLUTION folder.

On the disk it'll seem as they are in different folders, but in the solution there'll be a "Services" folder

mekk1t committed:
Removed blank spaces, moved ProjectService to "Services" folder (FINAL DECISION)
